### PR TITLE
webdriverio: middleware refactor

### PR DESCRIPTION
**Pull request by abjerstedt**
## Proposed changes

[//]: # (Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. If it fixes a bug or resolves a feature request, be sure to link to that issue.)

There is a current edge case with the most recent revision of the refetch handler:
expect($('selector').isDisplayed()).to.be.false 
^---- If the element falls stale between the fetch and the interaction, refetch was waiting on these elements when they weren't supposed to.

Extracting the implicit wait to its own file, and recycling this logic within the refetch.

## Types of changes

[//]: # (What types of changes does your code introduce to WebdriverIO?)
[//]: # (_Put an `x` in the boxes that apply_)

- [X] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

[//]: # (_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._)

- [X] I have read the [CONTRIBUTING](https://github.com/webdriverio/webdriverio/blob/master/CONTRIBUTING.md) doc
- [X] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

## Further comments

[//]: # (If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...)

### CODE REVIEW

File: packages/webdriverio/tests/commands/element/moveTo.test.js [hunk 1]
```diff
@@ -60,7 +60,7 @@ describe('moveTo', () => {
 
         const elem = await browser.$('#elem')
         request.setMockResponse([{}, {}])
-        expect(elem.moveTo(5, 10)).rejects.toThrow('Failed to getElementRect')
+        await expect(elem.moveTo(5, 10)).rejects.toThrow('Failed to receive element rects via execute command')
```
Comment (abjerstedt) [inline packages/webdriverio/tests/commands/element/moveTo.test.js]: This is currently broken due to the command not being awaited. This PR just exposes the broken UT.

File: packages/webdriverio/tests/commands/element/shadow$$.test.js [hunk 2]
```diff
@@ -9,15 +17,8 @@ describe('shadow$$', () => {
...
-        // 'Cannot assign to read only property'
-        const clone = Object.assign({}, el)
-        clone.shadow$$ = el.shadow$$
-        clone.$$ = el.$$
-        const spy = jest.spyOn(clone,'$$')
-        clone.shadow$$('#shadowfoo')
-        expect(spy).toHaveBeenCalledWith(expect.any(Function))
-        spy.mockClear()
+
+        await el.shadow$$('#shadowfoo')
```
Comment (abjerstedt) [inline packages/webdriverio/tests/commands/element/shadow$$.test.js]: This is currently broken due to the command not being awaited. This PR just exposes the broken UT (and cleans it up).

File: packages/webdriverio/tests/commands/element/shadow$.test.js [hunk 3]
```diff
@@ -9,15 +17,8 @@ describe('shadow$', () => {
...
-        // 'Cannot assign to read only property'
-        const clone = Object.assign({}, el)
-        clone.shadow$ = el.shadow$
-        clone.$ = el.$
-        const spy = jest.spyOn(clone,'$')
-        clone.shadow$('#shadowfoo')
-        expect(spy).toHaveBeenCalledWith(expect.any(Function))
-        spy.mockClear()
+
+        await el.shadow$('#shadowfoo')
```
Comment (abjerstedt) [inline packages/webdriverio/tests/commands/element/shadow$.test.js]: This is currently broken due to the command not being awaited. This PR just exposes the broken UT (and cleans it up).

File: packages/webdriverio/src/utils/implicitWait.js [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+             * if waitForExist was successful requery element and assign elementId to the scope
+             */
+            return await currentElement.parent.$(currentElement.selector)
+        } catch {
+            throw new Error(
+                `Can't call ${commandName} on element with selector "${currentElement.selector}" because element wasn't found`)
+        }
+    }
+
+    return currentElement
```
Comment (abjerstedt) [inline packages/webdriverio/src/utils/implicitWait.js]: Just fall through and return itself if not necessary.

File: packages/webdriverio/tests/commands/element/moveTo.test.js (hunk 1 above)
Comment (mgrybyk) [reply packages/webdriverio/tests/commands/element/moveTo.test.js]: looks like my fault

Comment (mgrybyk) [reply packages/webdriverio/tests/commands/element/moveTo.test.js]: @abjerstedt thanks for fixing this one!

File: packages/webdriverio/tests/commands/element/shadow$$.test.js [hunk 5]
```diff
@@ -1,5 +1,13 @@
 import { remote } from '../../../src'
 
+jest.mock('../../../src/commands/element/$$', () => ({
+    __esModule: true,
+    default: jest.fn().mockImplementation(() => { })
+}))
+
+const $$ = require('../../../src/commands/element/$$')
+
+
```
Comment (mgrybyk) [inline packages/webdriverio/tests/commands/element/shadow$$.test.js]: should we have here only one new line instead of two?

**Comment by mgrybyk** [comment]
@abjerstedt looks awesome! Have you tried these changes in real project?

File: packages/webdriverio/src/utils/implicitWait.js [hunk 6]
```diff
@@ -0,0 +1,30 @@
+import logger from '@wdio/logger'
+const log = logger('webdriverio')
+
+export default async function implicitWait (currentElement, commandName) {
+    /**
+     * wait on element if:
+     *  - elementId couldn't be fetched in the first place
+     *  - command is not explicit wait command for existance or displayedness
+     */
+
```
Comment (christian-bromann) [inline packages/webdriverio/src/utils/implicitWait.js]:
```suggestion
```

File: packages/webdriverio/tests/commands/element/shadow$$.test.js (hunk 5 above)
Comment (christian-bromann) [inline packages/webdriverio/tests/commands/element/shadow$$.test.js]:
```suggestion
```

File: packages/webdriverio/tests/commands/element/shadow$.test.js [hunk 7]
```diff
@@ -1,6 +1,14 @@
...
 
-describe('shadow$', () => {
+jest.mock('../../../src/commands/element/$', () => ({
+    __esModule: true,
+    default: jest.fn().mockImplementation(() => { })
+}))
+
+const $ = require('../../../src/commands/element/$')
+
+
```
Comment (christian-bromann) [inline packages/webdriverio/tests/commands/element/shadow$.test.js]:
```suggestion
```

File: packages/webdriverio/src/utils/implicitWait.js (hunk 6 above)
Comment (christian-bromann) [reply packages/webdriverio/src/utils/implicitWait.js]: Also the comment above can be moved on top of the function declaration

File: packages/webdriverio/tests/commands/element/shadow$$.test.js (hunk 5 above)
Comment (abjerstedt) [reply packages/webdriverio/tests/commands/element/shadow$$.test.js]: Fixed.

File: packages/webdriverio/tests/commands/element/shadow$.test.js (hunk 7 above)
Comment (abjerstedt) [reply packages/webdriverio/tests/commands/element/shadow$.test.js]: Fixed.

File: packages/webdriverio/tests/commands/element/shadow$$.test.js (hunk 5 above)
Comment (abjerstedt) [reply packages/webdriverio/tests/commands/element/shadow$$.test.js]: Fixed.

**Comment by abjerstedt** [comment]
I didn't mean to dismiss the reviews... not sure how I did that :-/. @WillBrock @christian-bromann 

@mgrybyk give me a few and I'll cross link my repos and smoke test this.

File: packages/webdriverio/src/utils/implicitWait.js (hunk 6 above)
Comment (abjerstedt) [reply packages/webdriverio/src/utils/implicitWait.js]: Fixed.

**Comment by abjerstedt** [comment]
@mgrybyk Tests look good (and i purposefully ran ones that i knew had stale elements)

**Comment by abjerstedt** [comment]
@christian-bromann @mgrybyk @WillBrock 
So I tweaked behavior a little bit more:
$(A).$(B).isDisplayed()
In the refetch scenario, $(A) will be waited for, but $(B) wouldn't.

**Pull request merged**